### ISSUE TITLE
OpenTelemetry Input Plugin and OpenTelemetry SDK Exporters - Fluentbit throws HTTP 400 Error

### ISSUE THREAD

**Issue by 0xjbushell** [fluent/fluent-bit#7678]
## Bug Report

**Describe the bug**
When using OpenTelemetry SDK Exporters and OpenTelemetry Input plugin, specifically http/protobuf. We're seeing the OpenTelemetry Input plugin / fluentbit return with an HTTP 400 Error "Invalid Response". 

After some debugging and packet captures, it appears that this is due to OpenTelemetry SDK Exporters sending the request with header: `Transfer-Encoding: chunked`. Looking through the source code, it looks like Fluentbit checks `Content-Length` header and if it is equal to 0, it will throw HTTP 400 / Invalid Response. https://github.com/fluent/fluent-bit/blob/master/lib/monkey/mk_server/mk_http_parser.c#L372-L379

However, it seems common/normal for OpenTelemetry to send requests with `Transfer-Encoding: Chunked`. I'm not exactly sure if this is a bug or not (seems like it could be) but would love some guidance / help on understanding the behavior. It seems like it *should* work out of the box as we are running very standard boiler plate / implementation of the OTEL SDK Exporters.

**To Reproduce**
Our use of OpenTelemetry SDK is pretty boilerplate, use OpenTelemetry SDK and send to OpenTelemetry Input Plugin.

- Example log message if applicable:
```
[2023/07/10 18:56:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/opentelemetry.c:52] new TCP connection arrived FD=45
[2023/07/10 18:56:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=0 now_len=1934
[2023/07/10 18:57:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=1934 now_len=3868
[2023/07/10 18:58:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=3868 now_len=5802
[2023/07/10 18:58:13] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=5802 now_len=7736
[2023/07/10 18:58:13] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:84] fd=45 closed connection
```
I took packet captures in order to get the response and understand details of the requests / responses. What I did to validate this further is export Portobuf payload, and send manually via curl in a few ways:

The first request just sends a captured protobuf payload (that was originally sent from OTEL Exporter). This seems to validate that the payload and contents itself are good.

```
curl -i -X POST http://${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" --data-binary @test-protobuf-curl-1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6979    HTTP/1.1 201 Created    0      0 --:--:-- --:--:-- --:--:--     0
0     0  100  6979      Server: Fluent Bit v2.1.5
Content-Length: 0

0   662k --:--:-- --:--:-- --:--:--  681k
```

This request takes the same payload but sets `Transfer-Encoding: chunked` like OTEL Exporter does, this fails with same error code / message.
```
curl -i -X POST http:/${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" -H "Transfer-Encoding: chunked" --data-binary @test-protobuf-curl-1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7002  100    23  100  6992   1286   382k --:--:-- --:--:-- --:--:--  402k
HTTP/1.1 400 Forbidden
Server: Fluent Bit v2.1.5
Content-Length: 23

error: invalid request
```

This request takes the same payload but sets `Transfer-Encoding` AND `Content-Length` in order to validate where things are failing. As shown above, it appears that fluentbit checks for `Content-Length` and if it is `==0`. it will throw the HTTP 400 / Invalid request.
```
curl -i -X POST http://${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" -H "Transfer-Encoding: chunked" -H "Content-Length: 6979" --data-binary @test-protobuf-curl-1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
      HTTP/1.1 201 Created
Server: Fluent Bit v2.1.5
Content-Length: 0

                           Dload  Upload   Total   Spent    Left  Speed
100  6979    0     0  100  6992      0   563k --:--:-- --:--:-- --:--:--  569k
```

- Steps to reproduce the problem:
- Use OpenTelemetry .NET SDK / Exporter for Traces / Metrics and configure for HTTPProtobuf
- Use OpenTelemetry Input Plugin
- Trigger Trace or Metric

**Expected behavior**
Fluentbit should not respond with an HTTP 400 / Invalid Request and attempt to route / process accordingly.

**Your Environment**

* Version used: v2.1.5
* Configuration: .NET / OpenTelemetry SDK
* Environment name and version (e.g. Kubernetes? What version?): Kubernetes / v1.25.6
* Server type and version:
* Operating System and version: Linux
* Filters and plugins: Nothing used for OpenTelmetry Plugin yet

**Additional context**

Cannot get an End-To-End working example with OpenTelemetry SDK and OpenTelemetry Input Plugin

**Comment by 0xjbushell** [comment]
Just wanted to add a comment that I also tested this using the boilerplate AspNetCore example app in the opentelemetry-dotnet repo: https://github.com/open-telemetry/opentelemetry-dotnet/tree/ad08228dfec8fd7d8839e09f68e9f0caef4233f4/examples/AspNetCore

I'm seeing the same result in requests. OTEL Exporter sets `Transfer-Encoding: chunked` which fluentbit responds with HTTP 400 / invalid request!

**Comment by cbi-at-varian** [comment]
Hi @0xjbushell, I think I might help with a setup to reproduce the issue. I stumbled over the same thing, but didn't get that far with the analysis. However #6441 is afais about the same and an easy setup to reproduce the issue can be found here https://github.com/cbi-at-varian/otlp-forwarding
I just updated it with the latest versions (fluentbit 2.1.7 and  .NET / OpenTelemetry SDK 1.5.1)

Would be nice to get that combo working!

**Comment by knaruf** [comment]
I have the same issue when trying to get traces from 
.NET to fluentbit. Would be very nice if this issue can be solved.

**Comment by knaruf** [comment]
I found a way to make it work in .NET:
```
services.AddHttpClient("OtlpTraceExporter")
    .ConfigureHttpMessageHandlerBuilder(builder =>
    {
        builder.AdditionalHandlers.Add(new ContentLengthDelegatingHandler());
    });
```
```
public class ContentLengthDelegatingHandler : DelegatingHandler
    {
        protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            request.Content!.LoadIntoBufferAsync().GetAwaiter().GetResult();
            return base.Send(request, cancellationToken);
        }

        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            await request.Content!.LoadIntoBufferAsync();
            return await base.SendAsync(request, cancellationToken);
        }
    }
```
Calling LoadIntoBufferAsync will cause the `Content-Length` header to be set instead of the `Transfer-Encoding: Chunked` header

**Comment by dceravigupta** [comment]
I'm facing the same issue when trying to get Envoy traces to FluentBit.

![image](https://github.com/fluent/fluent-bit/assets/32182026/d267bc4c-9d5b-4147-a38a-9cef29c42269)

**Comment by madhub** [comment]
> ## Bug Report
> 
> **Describe the bug** When using OpenTelemetry SDK Exporters and OpenTelemetry Input plugin, specifically http/protobuf. We're seeing the OpenTelemetry Input plugin / fluentbit return with an HTTP 400 Error "Invalid Response".
> 
> After some debugging and packet captures, it appears that this is due to OpenTelemetry SDK Exporters sending the request with header: `Transfer-Encoding: chunked`. Looking through the source code, it looks like Fluentbit checks `Content-Length` header and if it is equal to 0, it will throw HTTP 400 / Invalid Response. https://github.com/fluent/fluent-bit/blob/master/lib/monkey/mk_server/mk_http_parser.c#L372-L379
> 
> However, it seems common/normal for OpenTelemetry to send requests with `Transfer-Encoding: Chunked`. I'm not exactly sure if this is a bug or not (seems like it could be) but would love some guidance / help on understanding the behavior. It seems like it _should_ work out of the box as we are running very standard boiler plate / implementation of the OTEL SDK Exporters.
> 
> **To Reproduce** Our use of OpenTelemetry SDK is pretty boilerplate, use OpenTelemetry SDK and send to OpenTelemetry Input Plugin.
> 
>     * Example log message if applicable:
> 
> 
> ```
> [2023/07/10 18:56:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/opentelemetry.c:52] new TCP connection arrived FD=45
> [2023/07/10 18:56:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=0 now_len=1934
> [2023/07/10 18:57:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=1934 now_len=3868
> [2023/07/10 18:58:05] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=3868 now_len=5802
> [2023/07/10 18:58:13] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:89] read()=1934 pre_len=5802 now_len=7736
> [2023/07/10 18:58:13] [trace] [input:opentelemetry:input.opentelemetry at /src/fluent-bit/plugins/in_opentelemetry/http_conn.c:84] fd=45 closed connection
> ```
> 
> I took packet captures in order to get the response and understand details of the requests / responses. What I did to validate this further is export Portobuf payload, and send manually via curl in a few ways:
> 
> The first request just sends a captured protobuf payload (that was originally sent from OTEL Exporter). This seems to validate that the payload and contents itself are good.
> 
> ```
> curl -i -X POST http://${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" --data-binary @test-protobuf-curl-1
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100  6979    HTTP/1.1 201 Created    0      0 --:--:-- --:--:-- --:--:--     0
> 0     0  100  6979      Server: Fluent Bit v2.1.5
> Content-Length: 0
> 
> 0   662k --:--:-- --:--:-- --:--:--  681k
> ```
> 
> This request takes the same payload but sets `Transfer-Encoding: chunked` like OTEL Exporter does, this fails with same error code / message.
> 
> ```
> curl -i -X POST http:/${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" -H "Transfer-Encoding: chunked" --data-binary @test-protobuf-curl-1
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100  7002  100    23  100  6992   1286   382k --:--:-- --:--:-- --:--:--  402k
> HTTP/1.1 400 Forbidden
> Server: Fluent Bit v2.1.5
> Content-Length: 23
> 
> error: invalid request
> ```
> 
> This request takes the same payload but sets `Transfer-Encoding` AND `Content-Length` in order to validate where things are failing. As shown above, it appears that fluentbit checks for `Content-Length` and if it is `==0`. it will throw the HTTP 400 / Invalid request.
> 
> ```
> curl -i -X POST http://${ENDPOINT}.default.svc.cluster.local:4318/v1/traces -H "Content-Type: application/x-protobuf" -H "Transfer-Encoding: chunked" -H "Content-Length: 6979" --data-binary @test-protobuf-curl-1
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>       HTTP/1.1 201 Created
> Server: Fluent Bit v2.1.5
> Content-Length: 0
> 
>                            Dload  Upload   Total   Spent    Left  Speed
> 100  6979    0     0  100  6992      0   563k --:--:-- --:--:-- --:--:--  569k
> ```
> 
>     * Steps to reproduce the problem:
> 
>     * Use OpenTelemetry .NET SDK / Exporter for Traces / Metrics and configure for HTTPProtobuf
> 
>     * Use OpenTelemetry Input Plugin
> 
>     * Trigger Trace or Metric
> 
> 
> **Expected behavior** Fluentbit should not respond with an HTTP 400 / Invalid Request and attempt to route / process accordingly.
> 
> **Your Environment**
> 
>     * Version used: v2.1.5
> 
>     * Configuration: .NET / OpenTelemetry SDK
> 
>     * Environment name and version (e.g. Kubernetes? What version?): Kubernetes / v1.25.6
> 
>     * Server type and version:
> 
>     * Operating System and version: Linux
> 
>     * Filters and plugins: Nothing used for OpenTelmetry Plugin yet
> 
> 
> **Additional context**
> 
> Cannot get an End-To-End working example with OpenTelemetry SDK and OpenTelemetry Input Plugin

Also if you observe, HTTP Status string error message is wrong , instead of **400 BadRequest**, it says **400 Forbidden**
Here is the affected source code file [opentelemetry_prot.c](https://github.com/fluent/fluent-bit/blob/33ea42feda1953a32fe86f7eb136bbba42a49e8e/plugins/in_opentelemetry/opentelemetry_prot.c#L103)


![image](https://github.com/fluent/fluent-bit/assets/8907962/46c4d899-129f-4da0-bf94-baa7f3777819)

**Comment by madhub** [comment]
Also if you observe, HTTP Status string error message is wrong , instead of 400 BadRequest, it says 400 Forbidden
Here is the affected source code file [opentelemetry_prot.c](https://github.com/fluent/fluent-bit/blob/33ea42feda1953a32fe86f7eb136bbba42a49e8e/plugins/in_opentelemetry/opentelemetry_prot.c#L103)

![295872960-46c4d899-129f-4da0-bf94-baa7f3777819](https://github.com/fluent/fluent-bit/assets/8907962/b5f3f2ae-c970-41af-be7e-a5a74c465a6d)

**Comment by madhub** [comment]
> Also if you observe, HTTP Status string error message is wrong , instead of 400 BadRequest, it says 400 Forbidden Here is the affected source code file [opentelemetry_prot.c](https://github.com/fluent/fluent-bit/blob/33ea42feda1953a32fe86f7eb136bbba42a49e8e/plugins/in_opentelemetry/opentelemetry_prot.c#L103)
> 
> ![295872960-46c4d899-129f-4da0-bf94-baa7f3777819](https://private-user-images.githubusercontent.com/8907962/295874543-b5f3f2ae-c970-41af-be7e-a5a74c465a6d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMzAyNDEsIm5iZiI6MTcxMzMyOTk0MSwicGF0aCI6Ii84OTA3OTYyLzI5NTg3NDU0My1iNWYzZjJhZS1jOTcwLTQxYWYtYmU3ZS1hNWE3NGM0NjVhNmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTdUMDQ1OTAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzIxZDM1OGM5NGJlMWUxMGUxNTdlNmFiNGViNzIzODFjNTdhYmE5MmQ5ZDNmZmRkMmNjYTcwYTM3MThmMmQ4NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.6KzNOiTLKbIlf7ugzni6zkZZZm8fFUhy_XDGAlDa1s8)

Why this issue closed , its bug needs to be fixed

**Comment by loganrosen** [comment]
I haven't verified this, but it looks like this issue may have been fixed in #9466.
